registry/remote/auth/example_test.go

```diff
@@ -0,0 +1,219 @@
+		path := r.URL.Path
+		if r.Method != http.MethodGet {
+			w.WriteHeader(http.StatusNotFound)
+			panic("unexpected access")
+		}
+		switch path {
+		case "/basicAuth":
+			wantedAuthHeader := "Basic " + base64.StdEncoding.EncodeToString([]byte(username+":"+password))
+			authHeader := r.Header.Get("Authorization")
+			if authHeader != wantedAuthHeader {
+				w.Header().Set("Www-Authenticate", `Basic realm="Test Server"`)
+				w.WriteHeader(http.StatusUnauthorized)
+			}
+		case "/accessToken":
+			wantedAuthHeader := "Bearer " + accessToken
+			if auth := r.Header.Get("Authorization"); auth != wantedAuthHeader {
+				challenge := fmt.Sprintf("Bearer realm=%q,service=%q,scope=%q", as.URL, host, strings.Join(scopes, " "))
+				w.Header().Set("Www-Authenticate", challenge)
+				w.WriteHeader(http.StatusUnauthorized)
+			}
+		case "/clientConfig":
+			wantedAuthHeader := "Basic " + base64.StdEncoding.EncodeToString([]byte(username+":"+password))
+			authHeader := r.Header.Get("Authorization")
+			if authHeader != wantedAuthHeader {
+				w.Header().Set("Www-Authenticate", `Basic realm="Test Server"`)
+				w.WriteHeader(http.StatusUnauthorized)
+			}
+		case "/simple":
+			w.WriteHeader(http.StatusOK)
+		default:
+			w.WriteHeader(http.StatusNotAcceptable)
+		}
+	}))
+	defer ts.Close()
+	host = ts.URL
+	simpleURL = fmt.Sprintf("%s/simple", host)
+	basicAuthURL = fmt.Sprintf("%s/basicAuth", host)
+	accessTokenURL = fmt.Sprintf("%s/accessToken", host)
+	clientConfigURL = fmt.Sprintf("%s/clientConfig", host)
+	http.DefaultClient = ts.Client()
+
+	os.Exit(m.Run())
+}
+
+// ExampleClient_Do_minimalClient gives an example of a minimal working client.
+func ExampleClient_Do_minimalClient() {
+	var client auth.Client
+	req, err := http.NewRequest(http.MethodGet, simpleURL, nil)
+	if err != nil {
+		panic(err)
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Println(resp.StatusCode)
+	// Output:
+	// 200
+}
+
+// ExampleClient_Do_basicAuth gives an example of using client with credentials.
+func ExampleClient_Do_basicAuth() {
+	client := &auth.Client{
+		Credential: func(ctx context.Context, reg string) (auth.Credential, error) {
+			switch reg {
+			case host[8:]:
+				return auth.Credential{
+					Username: username,
+					Password: password,
+				}, nil
+			default:
+				return auth.EmptyCredential, fmt.Errorf("credential not found for %v", reg)
+			}
+		},
+	}
+	req, err := http.NewRequest(http.MethodGet, basicAuthURL, nil)
+	if err != nil {
+		panic(err)
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Println(resp.StatusCode)
+	// Output:
+	// 200
+}
+
+// ExampleClient_Do_withAccessToken gives an example of using client with an access token.
+func ExampleClient_Do_withAccessToken() {
+	client := &auth.Client{
+		Credential: func(ctx context.Context, reg string) (auth.Credential, error) {
+			return auth.Credential{
+				AccessToken: accessToken,
+			}, nil
```

Comment:
`reg` should always be checked before returning the credential

Reply:
Added reg checking here.